File: backend/src/main/java/com/apocalypse/thefall/service/inventory/factory/DefaultItemInstanceFactory.java
package com.apocalypse.thefall.service.inventory.factory;

import com.apocalypse.thefall.entity.instance.AmmoInstance;
import com.apocalypse.thefall.entity.instance.ArmorInstance;
import com.apocalypse.thefall.entity.instance.ItemInstance;
import com.apocalypse.thefall.entity.instance.WeaponInstance;
import com.apocalypse.thefall.entity.item.Ammo;
import com.apocalypse.thefall.entity.item.Armor;
import com.apocalypse.thefall.entity.item.Item;
import com.apocalypse.thefall.entity.item.Weapon;
import org.springframework.stereotype.Component;

@Component
public class DefaultItemInstanceFactory implements ItemInstanceFactory {

    @Override
    public ItemInstance createInstance(Item item) {
        if (item instanceof Weapon) {
            return WeaponInstance.builder()
                    .item(item)
                    .currentAmmoQuantity(0)
                    .build();
        } else if (item instanceof Armor) {
            return ArmorInstance.builder()
                    .item(item)
                    .build();
        } else if (item instanceof Ammo) {
            return AmmoInstance.builder()
                    .item(item)
                    .quantity(0)
                    .build();
        }
        throw new IllegalArgumentException("Unsupported item type: " + item.getClass().getSimpleName());
    }
}

File: backend/src/main/java/com/apocalypse/thefall/service/inventory/factory/DefaultItemInstanceFactory.kt
package com.apocalypse.thefall.service.inventory.factory

import com.apocalypse.thefall.entity.instance.AmmoInstance
import com.apocalypse.thefall.entity.instance.ArmorInstance
import com.apocalypse.thefall.entity.instance.ItemInstance
import com.apocalypse.thefall.entity.instance.WeaponInstance
import com.apocalypse.thefall.entity.item.Ammo
import com.apocalypse.thefall.entity.item.Armor
import com.apocalypse.thefall.entity.item.Item
import com.apocalypse.thefall.entity.item.Weapon
import org.springframework.stereotype.Component

@Component
class DefaultItemInstanceFactory : ItemInstanceFactory {

    override fun createInstance(item: Item): ItemInstance {
        return when (item) {
            is Weapon -> WeaponInstance().apply {
                this.item = item
                this.currentAmmoQuantity = 0
            }

            is Armor -> ArmorInstance().apply {
                this.item = item
            }

            is Ammo -> AmmoInstance().apply {
                this.item = item
                this.quantity = 0
            }

            else -> throw IllegalArgumentException("Unsupported item type: ${item::class.simpleName}")
        }
    }
}
